File: azdo-pbi-analyzer/src/utils/url-parser.ts
import { PBIMetadata } from '../models/pbi-data.js';

/**
 * Parses Azure DevOps PBI URLs and extracts metadata
 * Supports both dev.azure.com and visualstudio.com formats
 */

const URL_PATTERNS = [
    // https://dev.azure.com/{org}/{project}/_workitems/edit/{id}
    /https?:\/\/dev\.azure\.com\/([^/]+)\/([^/]+)\/_workitems\/edit\/(\d+)/,

    // https://{org}.visualstudio.com/{project}/_workitems/edit/{id}
    /https?:\/\/([^.]+)\.visualstudio\.com\/([^/]+)\/_workitems\/edit\/(\d+)/,

    // https://dev.azure.com/{org}/{project}/_workitems?id={id}
    /https?:\/\/dev\.azure\.com\/([^/]+)\/([^/]+)\/_workitems\?.*id=(\d+)/,
];

export class URLParser {
    /**
     * Parse Azure DevOps work item URL
     * @param url - Full Azure DevOps work item URL
     * @returns PBIMetadata object with organization, project, and work item ID
     * @throws Error if URL format is invalid
     */
    public static parse(url: string): PBIMetadata {
        for (const pattern of URL_PATTERNS) {
            const match = url.match(pattern);
            if (match) {
                const [, organization, project, workItemId] = match;
                return {
                    organization,
                    project: decodeURIComponent(project),
                    workItemId: parseInt(workItemId, 10),
                };
            }
        }

        throw new Error(
            `Invalid Azure DevOps URL format. Expected formats:\n` +
            `  - https://dev.azure.com/{org}/{project}/_workitems/edit/{id}\n` +
            `  - https://{org}.visualstudio.com/{project}/_workitems/edit/{id}\n` +
            `Received: ${url}`
        );
    }

    /**
     * Validate if URL is a valid Azure DevOps URL
     * @param url - URL to validate
     * @returns true if valid, false otherwise
     */
    public static isValid(url: string): boolean {
        try {
            this.parse(url);
            return true;
        } catch {
            return false;
        }
    }

    /**
     * Extract work item ID from URL
     * @param url - Azure DevOps URL
     * @returns Work item ID
     */
    public static extractWorkItemId(url: string): number {
        const metadata = this.parse(url);
        return metadata.workItemId;
    }
}

File: azdo-pbi-analyzer/src/AzDoPbiAnalyzer.Core/Utils/UrlParser.cs
using System;
using System.Text.RegularExpressions;
using AzDoPbiAnalyzer.Core.Models;

namespace AzDoPbiAnalyzer.Core.Utils;

public static class UrlParser
{
    public static PBIMetadata Parse(string url)
    {
        if (string.IsNullOrWhiteSpace(url))
            throw new ArgumentException("URL cannot be empty");

        // Format 1: https://dev.azure.com/{org}/{project}/_workitems/edit/{id}
        var devAzureMatch = Regex.Match(url, @"dev\.azure\.com/([^/]+)/([^/]+)/_workitems/edit/(\d+)", RegexOptions.IgnoreCase);
        if (devAzureMatch.Success)
        {
            return new PBIMetadata(
                devAzureMatch.Groups[1].Value,
                Uri.UnescapeDataString(devAzureMatch.Groups[2].Value),
                int.Parse(devAzureMatch.Groups[3].Value)
            );
        }

        // Format 2: https://{org}.visualstudio.com/{project}/_workitems/edit/{id}
        var visualStudioMatch = Regex.Match(url, @"([^.]+)\.visualstudio\.com/([^/]+)/_workitems/edit/(\d+)", RegexOptions.IgnoreCase);
        if (visualStudioMatch.Success)
        {
            return new PBIMetadata(
                visualStudioMatch.Groups[1].Value,
                Uri.UnescapeDataString(visualStudioMatch.Groups[2].Value),
                int.Parse(visualStudioMatch.Groups[3].Value)
            );
        }
        
        // Format 3: Query param style https://dev.azure.com/{org}/{project}/_workitems?id={id}
        var queryMatch = Regex.Match(url, @"dev\.azure\.com/([^/]+)/([^/]+)/_workitems\?id=(\d+)", RegexOptions.IgnoreCase);
        if (queryMatch.Success)
        {
            return new PBIMetadata(
                queryMatch.Groups[1].Value,
                Uri.UnescapeDataString(queryMatch.Groups[2].Value),
                int.Parse(queryMatch.Groups[3].Value)
            );
        }

        throw new ArgumentException("Invalid Azure DevOps URL format");
    }
}
